contracts/external/nexus/WrappedCoveredPrincipalToken.sol

```diff
@@ -0,0 +1,238 @@
+    ///            to the contract by the `msg.sender` to make execution successful.
+    /// @param  _amount Amount of covered position / wrapped token `msg.sender` wants to mint.
+    /// @param  _expiration Timestamp at which the derived tranche would get expired.
+    /// @param  _wrappedPosition Address of the Wrapped position which is used to derive the tranche.
+    function mint(
+        uint256 _amount,
+        uint256 _expiration,
+        address _wrappedPosition,
+        PermitData calldata _permitCallData
+    ) external override {
+        require(isAllowedWp(_wrappedPosition), "WFP:INVALID_WP");
+        address _tranche = address(
+            _deriveTranche(_wrappedPosition, _expiration)
+        );
+        _usePermitData(_tranche, _permitCallData);
+        // Only allow minting when the position get expired.
+        require(
+            ITranche(_tranche).unlockTimestamp() < block.timestamp,
+            "WFP:POSITION_NOT_EXPIRED"
+        );
+        // Assumed that msg.sender provides the sufficient approval the contract.
+        IERC20(_tranche).safeTransferFrom(
+            msg.sender,
+            address(this),
+            _fromWad(_amount, _tranche)
+        );
+        // Mint the corresponding wrapped token to the `msg.sender`.
+        _mint(msg.sender, _amount);
+    }
+
+    /// @notice Tell whether the given `_wrappedPosition` is whitelisted or not.
+    /// @param  _wrappedPosition Address of the wrapped position.
+    /// @return returns boolean, True -> allowed otherwise false.
+    function isAllowedWp(address _wrappedPosition)
+        public
+        view
+        override
+        returns (bool)
+    {
+        return _allowedWrappedPositions.contains(_wrappedPosition);
+    }
+
+    /// @notice Returns the list of wrapped positions that are whitelisted with the contract.
+    ///         Order is not maintained.
+    /// @return Array of addresses.
+    function allWrappedPositions()
+        external
+        view
+        override
+        returns (address[] memory)
+    {
+        return _allowedWrappedPositions.values();
+    }
+
+    /// @notice Reclaim tranche token (i.e principal token) by the authorized account.
+    /// @dev    Only be called by the address which has the `RECLAIM_ROLE`, Should be Nexus Treasury.
+    /// @param  _expiration Timestamp at which the derived tranche would get expired.
+    /// @param  _wrappedPosition Address of the Wrapped position which is used to derive the tranche.
+    function reclaimPt(uint256 _expiration, address _wrappedPosition)
+        external
+        override
+        onlyRole(RECLAIM_ROLE)
+    {
+        require(isAllowedWp(_wrappedPosition), "WFP:INVALID_WP");
+        address _tranche = address(
+            _deriveTranche(_wrappedPosition, _expiration)
+        );
+        uint256 amount = IERC20(_tranche).balanceOf(address(this));
+        IERC20(_tranche).safeTransfer(msg.sender, amount);
```

Comment:
Let's make the place where the funds are transferred a calldata arg instead of assuming it to be the sender.